### ISSUE TITLE
Feature Request: Add BCELossWithLogits to replace BCELoss for numerical stability.

### ISSUE THREAD

**Issue by yzgao** [pytorch/pytorch#751]
Hi, 

I recently switched my codebase from tensorflow to pytorch. After watching the tutorials as well as some example projects, I felt quite excited about pytorch. It is way more easier to use and more transparent than tensorflow. The source code is very enjoyable to read. Thanks for making it public.

I have a feature request stated as follows.

I am using BCELoss for training discriminator in GANs. However, I noticed that BCELoss is built on top of sigmoid layer, which is numerically unstable. In my tensorflow implementation, I used sigmoid_cross_entropy_with_logits, which can prevent the overflow situation that often happens when we start to train from scratch. 

I wrote a very basic module to replace BCELoss in my code. The code below illustrates the idea, and is definitely not a very formal code. I feel that it is necessary to add similar loss modules to address this numerical stability issue.

```
class StableBCELoss(nn.modules.Module):
       def __init__(self):
             super(StableBCELoss, self).__init__()
       def forward(self, input, target):
             neg_abs = - input.abs()
             loss = input.clamp(min=0) - input * target + (1 + neg_abs.exp()).log()
             return loss.mean()
```

**Comment by soumith** [comment]
thanks, it's been on my list of TODO. Will make BCELoss out of the box have stable gradient.

**Comment by soumith** [comment]
@colesbury this is the task that you are interested in :P it's even implemented.

**Comment by alykhantejani** [comment]
@soumith @colesbury I can send a PR for this if no one is working on it? 

Currently `BCELoss` assumes a sigmoid is applied before this layer, so do we want to add a new layer all together for the more stable version that comibne `SoftMax` + `BCELoss`? If so, any suggestions for a name?

**Comment by soumith** [comment]
don't have a good name for this, but please go for it :)

**Comment by alykhantejani** [comment]
@soumith ok great, I'll come up with something. So the question is, should I implement it in C + CUDA or just in Python? I know BCELoss was originally in Lua and then moved to C. 

I'm happy to do it the C and CUDA way if this add's any extra value? Let me know which way would be preferred.

**Comment by soumith** [comment]
python seems okay to begin with

**Comment by alykhantejani** [comment]
I think this issue can be closed now?

**Comment by spurra** [comment]
@yzgao I am having trouble understanding how you came up with this formulation:
1) How is this equivalent to BCE?
2) Where do the numerical instabilities arise in BCE and how does this formulation fix it?

Thank you for your time!

**Comment by piantou** [comment]
> @yzgao I am having trouble understanding how you came up with this formulation:
> 
> 1. How is this equivalent to BCE?
> 2. Where do the numerical instabilities arise in BCE and how does this formulation fix it?
> 
> Thank you for your time!

you can see it at http://gameofdimension.com/2017/10/29/numerical-stability/
